### ISSUE TITLE
[makefilename] issues in the help file

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#549]
help file says: "%f decimal floating point (currently broken)", but it isn't (it works great)

and we also have "%o" as "signed octal", but it's actually unsigned, right?

**Comment by porres** [comment]
found more missing patterns, see screenshot and test file. Note that "%li" is not working as an unsigned long integer.
[sprintf-test.pd.zip](https://github.com/pure-data/pure-data/files/2818459/sprintf-test.pd.zip)

<img width="702" alt="screen shot 2019-01-31 at 15 55 30" src="https://user-images.githubusercontent.com/5360142/52074323-bda85000-2570-11e9-8bb6-64f1c8dbee6a.png">

**Comment by porres** [comment]
I'm sure there are even more undocumented patterns, and, anyway, we have to check it out and see which are desired or not... I'm on macOS Mojave, using 0.49-1, BTW

**Comment by porres** [comment]
I think at least %g and %G should be officially included and documented as so...

**Comment by umlaeute** [comment]
it probably should just mention [printf(3)](https://linux.die.net/man/3/printf)

**Comment by porres** [comment]
but I don't believe all from that reference is implemented, available or that it would be intentional to have more than what was documented

**Comment by umlaeute** [comment]
> i don't believe that [...] it would be intentional to have more than what was documented

what do you mean with that?

**Comment by porres** [comment]
I mean that the documentation has a limited set of possible patterns that doesn't include all of what your reference presents (which is the complete set). This seems to intentionally limit the possible patterns. For instance: "%a" does not work in [makefilename]. So I guess @millerpuckette needs to clarify his intention.

**Comment by Spacechild1** [comment]
%a is indeed missing in the code. note that `%a` is only available in C99 (https://stackoverflow.com/questions/4826842/the-format-specifier-a-for-printf-in-c) and I think Pd is compiled with C89/90 (at least I remember one of my commits failing the CI because I accidentally used C99 features). maybe it's time to switch to a more recent C version :-)

**Comment by porres** [comment]
%F is also missingm but it;s just the same as "%f"

**Comment by umlaeute** [comment]
@Spacechild1 Pd doesn't do anything to force the compiler to C99. that's why it still compiles with compilers that don't support it (like the one in the CI).

i think this is a good thing (at least not something i would trade for "%F" or "variable declarations all over the place")

**Comment by Spacechild1** [comment]
> like the one in the CI

ugh, I forgot about MSVC... although it seems that with VS 2013 (!) they finally implemented some C99 features. pretty ridiculous IMHO. <\rant>

but I agree that `%a` alone is not worth breaking backwards compatibility.

**Comment by Spacechild1** [comment]
> it probably should just mention printf(3)

to avoid any confusion, I think the help file should say that it implements a subset of `printf `and then list the available specifiers as found in the source code  @porres do you want to do that? ;-)

**Comment by porres** [comment]
Yeah, I could do that, wrapping up, documentation should:

- remove "currently broken" from %f
- say "**un**signed octal" instead of "signed"
- add "%g" and "%G"

Moreover, we should also mention the modifiers "l" that also seem to work, but, they also seem buggy as I mentioned before.

**Comment by porres** [comment]
here's what we got now for the documentation, which has been already merged...
 
![Screen Shot 2019-12-08 at 15 46 41](https://user-images.githubusercontent.com/5360142/70394235-fa837580-19d1-11ea-95e0-1d23d5d08234.png)

But there still seems to be issues, like https://github.com/pure-data/pure-data/issues/737 , which is a proper documentation of %p, I also left out the modifiers that exist and also seem to have trouble... ( see https://github.com/pure-data/pure-data/issues/549#issuecomment-459442198 )

I didn't mention either the %a pattern which is in the code. Let's try to reach a conclusion on how to address these in the documentation so we can close this issue or propose fixes.

**Comment by porres** [comment]
since https://github.com/pure-data/pure-data/issues/812#issuecomment-562975413 may eventually bring us some change in the code, we can enjoy this opportunity and also wrap this up alongside

**Comment by Spacechild1** [comment]
> I didn't mention either the %a pattern which is in the code

`%a` is not in the code. The documentation for `%p` should be fixed to say something like "platform dependent pointer representation". The `l` and `ll` specifiers are not honored (yet).

**Comment by porres** [comment]
Ok, so all we should bother so far about the issues in documentation is just '%p" and leave the rest as it is, huh?

**Comment by porres** [comment]
Well, here's the deal, the '%p' option wasn't in the documentation in 0.48 - now, I don't remember who changed the documentation and how it got in 0.49. I have the idea I included more information, or at least edit it, but I couldn't find the changes in the repository history.

So, where I'm getting at is this, is the '%p' option even desirable? It doesn't seem useful to have an option for "platform dependent pointer representation", right? So maybe just don't mention it? Remove it?

**Comment by porres** [comment]
is there a good way to search for file changes on github? Like when this help file changed from 0.48 to 0.49?

**Comment by Spacechild1** [comment]
> is there a good way to search for file changes on github?

View the file in GitHub, then click on the line number, click on "View git blame". In your case you'll see something like this and find the commit which introduced the change: https://github.com/pure-data/pure-data/blame/3c12b7694df6e4694aa57487fe8f7256a75677ca/doc/5.reference/makefilename-help.pd#L1

**Comment by Spacechild1** [comment]
> So, where I'm getting at is this, is the '%p' option even desirable? I

I've never needed it, but in the code there's a dedicated `POINTER` member in `t_printtype`, so `%p` has been implemented on purpose and should therefore be documented.

**Comment by porres** [comment]
> View the file in GitHub, then click on the line number, click on "View git blame". In your case you'll see something like this and find the commit which introduced the change: https://github.com/pure-data/pure-data/blame/3c12b7694df6e4694aa57487fe8f7256a75677ca/doc/5.reference/makefilename-help.pd#L1

Thanks, so now I an see @megrimm did change the documentation and put 'unsigned hexadecimal in lowercase preceeded by "0x"' in there, see https://github.com/pure-data/pure-data/commit/18c393c1b9305a72fefbca3a33adc1eaefa21565

I did work on the help file later, but I wasn't the one who put this there - I was wondering if Miller had put this there and for a moment I believed I was the one to blame.

It'd be great if @millerpuckette would tell us why he thought it was useful to include this in the code. I have no idea and I think it's supposed to be a way to format pointers into symbols? And it would also be good to know if he's aware there's a platform discrepancy here, so different results are given for Windows/Mac/Linux.

For the momento maybe I can just say it's "pointer representation" and follow it by "(platform dependent)", like that, inside parenthesis, and then we can see later if that platform dependancy is a issue we could further address.

**Comment by porres** [comment]
>  I think it's supposed to be a way to format pointers into symbols?

no, it gives us: [makefilename: no method for 'pointer']

**Comment by umlaeute** [comment]
@porres what makes you think that miller "included that code"? (apart from him being the gatekeeper, of everything that goes into Pd).

have you ever wanted to print the address of a string (or anything else)? I did.

**Comment by porres** [comment]
> what makes you think that miller "included that code"? 

I didn't say that, I was just wondering who put that description of %p in the help file from 0.48 to 0.49, which was wrong, so it seems... I considered it could have been miller.

> have you ever wanted to print the address of a string (or anything else)?

nope, I can't even say I know what that means, let alone why I'd need it - so I'm asking you people about it before I put something not quite accurate in the documentation. Anyway, I'll go ahead and change this as I mentioned : "pointer representation (platform dependent)", hold on.

**Comment by umlaeute** [comment]
thanks for updating the documentation.

could you please also rename the subpatch (and title) "*all possible patterns*" to just "**possible patterns**"? the help-patch completely misses `field width` and `flag` modifiers, and they make it practically impossible to give something like a *complete* list of patterns.

**Comment by porres** [comment]
ok

**Comment by porres** [comment]
Hi, as long as we're updating the documentation and my proposal hasn't been reviewed or included yet by Miller, I'm thinking of adding examples for all documented formats in the help patch.

I'm still quite clueless about the '%p' format. I was just specifying it was " "pointer representation (platform dependent)" as discussed here. But I'm not sure how it works and how it differs from one platform to another.

I now realize it can also receive symbols and generate a result accordingly, but I don't know what to write about that...

**Comment by porres** [comment]
> have you ever wanted to print the address of a string (or anything else)? I did.

@umlaeute , could you give me a use case/example? maybe it'd be nice to mention it in the documentation.

**Comment by Spacechild1** [comment]
> "pointer representation (platform dependent)"

that's perfectly fine. the main use is for debugging, I guess. if the user doesn't know what the sentence above means, they ain't gonna need it :-)

**Comment by Spacechild1** [comment]
> But I'm not sure how it works and how it differs from one platform to another.

https://stackoverflow.com/questions/2369541/where-is-p-useful-with-printf

**Comment by porres** [comment]
>  https://stackoverflow.com/questions/2369541/where-is-p-useful-with-printf

what I get from this is that you cannot depend on %p displaying a 0x prefix. s it says somewhere there "On Visual C++, it does not. Use %#p to be portable."

well, I tested on windows and in fact I didn't get 0x, but nonetheless, it's a hexadecimal representation.

Anyway, as far as documentation goes in my examples, how's this?

<img width="483" alt="Screen Shot 2020-05-02 at 12 23 59" src="https://user-images.githubusercontent.com/5360142/80868266-d1f56400-8c6f-11ea-937c-27fab3dcfa1a.png">

I already have a new documentation file with all examples, I'm just wondering about this one, and I  can make the commit if you think this is fine...

**Comment by Spacechild1** [comment]
I don't think you should mention `void *` here. Also, the representation doesn't have to be (always) hexidecimal. Generally it is, but some implementations print something like "NULL" for a null pointer. 

Just say that it gives you a platform specific pointer representation for a float or symbol.

Actually, I don't even know what's the purpose of sending a float to `%p`, it will just cast it to an int and then treat is as a pointer address, which seems kind of point(er)less to me.

Ironically, `%p` doesn't accept gpointers, but that's another issue.

**Comment by porres** [comment]
Ok, this is what I have now...

<img width="474" alt="Screen Shot 2020-05-02 at 12 55 18" src="https://user-images.githubusercontent.com/5360142/80869003-3286a000-8c74-11ea-8be9-06dcbc4f9fef.png">

> Ironically, %p doesn't accept gpointers, but that's another issue.

I thought about that. It should, right?

> Actually, I don't even know what's the purpose of sending a float to %p, it will just cast it to an int and then treat is as a pointer address, which seems kind of point(er)less to me.

Seems then I should I remove the float from the example? 

I could then just write something like "of an incoming symbol (floats are cast to int and also converted)"

**Comment by Spacechild1** [comment]
Your picture looks good to me.

> It should, right?

I think so. It seems more useful to me than feeding a float.

> I could then just write something like "of an incoming symbol (floats are cast to int and also converted)"

Maybe: "(floats are interpreted as memory addresses)".

The more I think of it, the more useless it seems to send a float to `%p`: they can't accurately represent memory addresses larger than 2^24 anway

**Comment by porres** [comment]
> The more I think of it, the more useless it seems to send a float to %p: they can't accurately represent memory addresses larger than 2^24 anway

I'm removing from the example then, sounds like pointless to point(er) it out.

<img width="445" alt="Screen Shot 2020-05-02 at 13 22 19" src="https://user-images.githubusercontent.com/5360142/80869634-f9502f00-8c77-11ea-9bcf-feb7b1110796.png">

anyway, rewrite is almost done, I'll make the commit so you all can revise it.

and maybe i'll open an issue to also allow "gpointers"

**Comment by porres** [comment]
Please check https://github.com/pure-data/pure-data/pull/829/commits/1e2db7633e11ba9ea58955e77f7dc55f139a48e5 with my additions of examples

**Comment by porres** [comment]
Ok, to wrap things up. I left the "length field" undocumented - reference: https://en.wikipedia.org/wiki/Printf_format_string#Length_field

I started this thread by pointing how some were not behaving properly.

@Spacechild1  said:

> The l and ll specifiers are not honored (yet).

I actually don't understand why, since it can be inserted and affects the output, but I just did not mention it, neither I mentioned it is not "honored" yet, just silently ignored.

By the way, there are other possibilities from https://en.wikipedia.org/wiki/Printf_format_string that are not implemented and thus ignored...

And on the other hand, there are some things that I don't know if are intentionally "honored" or expected, but it just works, like the usage of "+" and "#" flags.

Well, not entirely, cause at least  one thing doesn't really work as documented in https://en.wikipedia.org/wiki/Printf_format_string 

For instance, the usage of "#" for floats (f, F, e, E, g, G) say that if forces the output of the output always contains a decimal point, but that already happens, so I don't know if this should be treated as a bug and fixed...

Anyway, while we're at it, there's also a related bug that could be fixed, see: https://github.com/pure-data/pure-data/issues/812
